[jit] fix optional type subtype relation

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#22186 [jit] fix optional type subtype relation**

Differential Revision: [D15994644](https://our.internmc.facebook.com/intern/diff/D15994644)

wanchaol committed:
[jit] fix optional type subtype relation

wanchaol committed:
Update on "[jit] fix optional type subtype relation"

[jit] fix optional type subtype relation

gh-metadata: pytorch pytorch 22186 gh/wanchaol/26/head

nairbv commented on aten/src/ATen/core/jit_type.h:
```diff
@@ -259,9 +259,19 @@ struct CAFFE2_API OptionalType: public SingleElementType<TypeKind::OptionalType,
   bool operator==(const Type& rhs) const override {
     if (auto opt_rhs = rhs.cast<OptionalType>()) {
-      return getElementType()->isSubtypeOf(opt_rhs->getElementType());
+      return getElementType() == opt_rhs->getElementType();
```
ah I was looking at this and thinking the operator looked wrong. I tried changing it to this locally but it didn't fix my issue -- didn't realize we'd need the new isSubtypeOf as well to get it to work.

wanchaol replied:
Yup I think the `isSubtypeOf` will fix your problem

wanchaol committed:
Update on "[jit] fix optional type subtype relation"

[jit] fix optional type subtype relation

gh-metadata: pytorch pytorch 22186 gh/wanchaol/26/head

wanchaol committed:
remove equals as singleelementtype contains the overload on "[jit] fix optional type subtype relation"

[jit] fix optional type subtype relation

gh-metadata: pytorch pytorch 22186 gh/wanchaol/26/head
